tests/replication/test_sharded_event_persister.py

```diff
@@ -100,3 +103,171 @@ def test_basic(self):
 
         self.assertTrue(persisted_on_1)
         self.assertTrue(persisted_on_2)
+
+    def test_vector_clock_token(self):
+        """Tests that using a stream token with a vector clock component works
+        correctly with basic /sync and /messages usage.
+        """
+
+        self.make_worker_hs(
+            "synapse.app.generic_worker", {"worker_name": "worker1"},
+        )
+
+        self.make_worker_hs(
+            "synapse.app.generic_worker", {"worker_name": "worker2"},
+        )
+
+        sync_hs = self.make_worker_hs(
+            "synapse.app.generic_worker", {"worker_name": "sync"},
+        )
+
+        # Specially selected room IDs that get persisted on different workers.
+        room_id1 = "!foo:test"
+        room_id2 = "!baz:test"
+
+        self.assertEqual(
+            self.hs.config.worker.events_shard_config.get_instance(room_id1), "worker1"
+        )
+        self.assertEqual(
+            self.hs.config.worker.events_shard_config.get_instance(room_id2), "worker2"
+        )
+
+        user_id = self.register_user("user", "pass")
+        access_token = self.login("user", "pass")
+        requester = create_requester(user_id, access_token)
+
+        store = self.hs.get_datastore()
+
+        # Create two room on the different workers.
+        room_creator = self.hs.get_room_creation_handler()
+        self.get_success(
+            room_creator.create_room(requester, {}, requested_room_id=room_id1), by=0.1
```

Comment:
I'm surprised that the `by` params are needed here

Reply:
I'm not sure why, but we seem to need it whenever we send HTTP requests (e.g. the `render` methods automatically advance the clock repeatedly until a result comes in), which happens here where it sends HTTP replication requests to send events.